### ISSUE TITLE
Check on compile times

### ISSUE THREAD

**Issue by Mm2PL** [Chatterino/chatterino2#3452]
### Checklist

- [X] I'm reporting a problem with Chatterino
- [X] I've verified that I'm running **the most** recent nightly build or stable release
- [X] I've looked for my problem on the [wiki](https://wiki.chatterino.com/Help/)
- [X] I've searched the [issues and pull requests](https://github.com/Chatterino/chatterino2/issues?q=) for similar looking reports

### Describe your issue

Guys, we've got an issue in the form of compile times. It takes ~7 minutes for CI (which has decent computing power) to compile Chatterino. While I know that C++ itself has long compile times, it probably is possible to shorten the compile times at least a little.

### Screenshots

n/a

### OS and Chatterino Version

Since a long time ago xd

**Comment by karl-police** [comment]
I compiled things and it only compiled the files that were modified, because of like, idk cache or it checked what changed or something.

With AppVeyor it would have taken longer, because it compiles all the files again.

**Comment by Mm2PL** [comment]
> With AppVeyor it would have taken longer, because it compiles all the files again.

Yes. The compile times I'm talking about in this issue are compiling everything from scratch, exactly like AppVeyor. Incremental builds are a lot faster.

**Comment by karl-police** [comment]
> > With AppVeyor it would have taken longer, because it compiles all the files again.
> 
> Yes. The compile times I'm talking about in this issue are compiling everything from scratch, exactly like AppVeyor. Incremental builds are a lot faster.

I'm compiling a branch from scratch now on AppVeyor to see how long it takes there.

**Comment by karl-police** [comment]
> > > With AppVeyor it would have taken longer, because it compiles all the files again.
> > 
> > 
> > Yes. The compile times I'm talking about in this issue are compiling everything from scratch, exactly like AppVeyor. Incremental builds are a lot faster.
> 
> I'm compiling a Release version from a branch from scratch now on AppVeyor to see how long it takes there.

So a later build took 6 min 23 sec until it said "Build success" and the one I just made took 7 min 28 sec

**Comment by karl-police** [comment]
> > > > With AppVeyor it would have taken longer, because it compiles all the files again.
> > > 
> > > 
> > > Yes. The compile times I'm talking about in this issue are compiling everything from scratch, exactly like AppVeyor. Incremental builds are a lot faster.
> > 
> > 
> > I'm compiling a Release version from a branch from scratch now on AppVeyor to see how long it takes there.
> 
> So a later build took 6 min 23 sec until it said "Build success", and another later build took 6 min 46 sec
> 
> The one I just made took 7 min 28 sec

I ran right another one after the last one finished and it took 7 min 38 sec

**Comment by karl-police** [comment]
> Reference in n

and another one I just ran took 7 min 36 sec

**Comment by jammehcow** [comment]
I'm sure there's a bit of work that could be done on c2's side to improve compile times, but nothing's gonna get you around the specs of a GH runner unless you use a self-hosted one. Might be worth looking into.

**Comment by pajlada** [comment]
From the duplicate issue:

> Clangs --ftime-trace is useful for figuring out ways to reduce compile time
> 
> Otherwise upgrading the build infra to more cores

**Comment by Nerixyz** [comment]
> Clangs --ftime-trace is useful for figuring out ways to reduce compile time

On Windows for MSVC, you can use [microsoft/vcperf](https://github.com/microsoft/vcperf) and the [Windows Performance Analyzer](https://learn.microsoft.com/windows-hardware/get-started/adk-install) (about 700 MiB).

<details><summary>Building from CLI</summary>
<p>

To compile, you need to have [`nuget`](https://nuget.org/downloads) installed. Open a terminal with Visual Studio environment variables, and run the following commands:

```cmd
git clone https://github.com/microsoft/vcperf.git
cd vcperf
nuget restore
msbuild /property:Configuration=Release
```

The resulting `vcperf.exe` will be in `out/Release/x64`.

</p>
</details>

**Comment by leon-richardt** [comment]
I realize this may not be _super_ helpful but here's a link to an article by one of the Blender devs on common techniques to improve C++ build times: https://devtalk.blender.org/t/speed-up-c-compilation/30508
